monai/apps/auto3dseg/hpo_wrap.py

```diff
@@ -0,0 +1,152 @@
+# you may not use this file except in compliance with the License.
+# You may obtain a copy of the License at
+#     http://www.apache.org/licenses/LICENSE-2.0
+# Unless required by applicable law or agreed to in writing, software
+# distributed under the License is distributed on an "AS IS" BASIS,
+# WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+# See the License for the specific language governing permissions and
+# limitations under the License.
+
+import os
+import pickle
+import sys
+from abc import abstractmethod
+from copy import deepcopy
+
+import nni
+
+from monai.auto3dseg.algo_gen import AlgoGen
+from monai.bundle.config_parser import ConfigParser
+
+__all__ = ["HpoWrap", "NniWrapper"]
+
+
+class HpoWrap(AlgoGen):
+    """
+    This class generates a set of al, each of them can run independently.
+    """
+
+    @abstractmethod
+    def _get_hyperparameters():
+        """Get the hyperparameter from HPO"""
+        raise NotImplementedError("")
+
+    @abstractmethod
+    def _update_model():
+        """Update model params"""
+        raise NotImplementedError("")
+
+    @abstractmethod
+    def _report_results():
+        """Report result to HPO"""
+        raise NotImplementedError("")
+
+    @abstractmethod
+    def run_algo(self, obj_file, template_path):
+        """Interface for the HPO to run the training"""
+        raise NotImplementedError("")
+
+
+default_nni_config = {
+    "experimentName": "dummy",
+    "searchSpace": {"lr": {"_type": "choice", "_value": [0.0001, 0.001, 0.01, 0.1]}},
+    "trialCommand": "",  # will be replaced in the wrapper
+    "trialCommand": "python -m monai.apps.auto3dseg NniWrapper run_algo ",
+    "trialCodeDirectory": "",  # will be replaced in the wrapper
+    "trialGpuNumber": 1,
+    "trialConcurrency": 1,
+    "maxTrialNumber": 10,
+    "maxExperimentDuration": "1h",
+    "tuner": {"name": "GridSearch"},
+    "trainingService": {"platform": "local", "useActiveGpu": True},
+}
+
+
+class NniWrapper(HpoWrap):
+    """
+    Wrapper for NNI
+    """
+
+    def __init__(self, algo=None, **override):
+        """
+        Args:
+            algo: an dict that has {name: Algo object} format.
+                The Algo object must have get_script_path method
+            override:
+                a set of parameter to override the HPO config, e.g. search space
+        """
+        self.algo = algo
+        self.cfg = deepcopy(default_nni_config)
+        for k, v in override.items():
+            self.cfg.update({k: v})
+
+        if self.algo is not None:
+            if len(algo.keys()) > 1:
+                raise ValueError(f"object {algo} only allows 1 key, but there are {len(algo.keys())}")
+            name = list(algo.keys())[0]  # the only key
+            self.cfg.update({"experimentName": name})
+
+            # template path, fixed dir structure
+            template_path = os.path.abspath(os.path.join(algo[name].get_script_path(), "..", ".."))
```

Comment:
I found it not straightforward to instantiate the Algo/BundleAlgo class. Maybe we can add more information about the script path in the history from BundleGen @wyli ?

Reply:
sure, perhaps you can store `obj_file` filename in the history? looks like the obj file is maintained by this class.

Reply:
Yes, the history may contain the `obj_file` and a path to find the target (e.g. `segresnet2d.scripts.algo.Segresnet2dAlgo`)

Reply:
I found the package path might be very confusing, perhaps we can improve the error handling:

```py
try:
   self.algos.append(ConfigParser(algo_params).get_parsed_content())
except RuntimeError as e:
   if "ModuleNotFoundError" in str(e):
       msg = """Please make sure the folder structure of an Algo Template follows 
        [algo_name]
        ├── configs
        │   ├── hyperparameters.yaml  # automatically generated yaml from a set of ``template_configs``
        │   ├── network.yaml  # automatically generated network yaml from a set of ``template_configs``
        │   ├── transforms_train.yaml  # automatically generated yaml to define tranforms for training
        │   ├── transforms_validate.yaml  # automatically generated yaml to define transforms for validation
        │   └── transforms_infer.yaml  # automatically generated yaml to define transforms for inference
        └── scripts
            ├── test.py
            ├── __init__.py
            └── validate.py
       """
       raise RuntimeError(msg)
```